### PR TITLE
Compare our steady state detection to Georges et. al. (2007)

### PR DESCRIPTION
**Pull request by snim2**
Example output: [georges.pdf](https://github.com/softdevteam/warmup_experiment/files/914792/georges.pdf)

### CODE REVIEW

**Review by ltratt — APPROVED** [review]

File: bin/table_georges_comparison [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+def _start_table(format_, headings):
+    """Unlike the start_table() function in warmup.latex, this one does not
+    contain any code to configure sparklines.
+    """
+
+    return _START_TABLE(format_, headings)
+
+
+def _cov(data):
+    """Computes the Coefficient of Variation of data."""
```
Comment (ltratt) [inline bin/table_georges_comparison]: Small 'c' in "coefficient"?

File: bin/table_georges_comparison [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+                print('WARNING: Skipping: %s from %s (benchmark crashed)' %
+                      (key, machine))
+            else:
+                bench, vm, variant = key.split(':')
+                n_pexecs = len(data_dictionaries[machine]['wallclock_times'][key])
+                for pexec in xrange(n_pexecs):
+                    stop = len(data_dictionaries[machine]['wallclock_times'][key][pexec]) - steady
+                    classification = data_dictionaries[machine]['classifications'][key][pexec]
+                    counts[machine]['total ' + classification] += 1
+                    for iteration in xrange(len(data_dictionaries[machine]['wallclock_times'][key][pexec])):
```
Comment (ltratt) [inline bin/table_georges_comparison:73]: We need to think a bit about this, as we're imposing our idea of when the steady state might be onto Georges. For example, if there are 2000 iterations, we say a steady state must have happened by 1500. It *could* be argued that Georges might still find a steady state somewhere between 1500 and 2000. My initial inclination is to give them the benefit of the doubt. Any thoughts?

Comment (snim2) [reply bin/table_georges_comparison:73]: Georges also tries to find a steady state before a particular iteration number, so in this script I have retrieved the number we used (500) from the JSON and used the same number for the Georges method -- that is defined on line 70. In their paper they call our "steady" "k": "Suppose that we want to retain k measurements per invocation. ..."

**Comment by snim2** [comment]
This is the relevant part of the paper:

screenshot from 2017-04-12 10-25-45

**Comment by ltratt** [comment]
Yes, but as I read the code above, we stop at 1500 iterations and say "there is no steady state in our methodology, so Georges didn't find one either". Warning: I might be wrong.

**Comment by snim2** [comment]
I'm not sure I follow. In the code we are (hopefully) implementing the method from Georges. In Step 1 we say "we want to retain k=500 iterations". In Step 2 we say "loop over iterations 1 to 1501 and test whether `_cov(wallclock_times[iteration:]) < threshold`, if it is then Georges says that is a steady state".

**Comment by ltratt** [comment]
But what if Georges says it's a steady state after iteration 1500?

**Comment by snim2** [comment]
Like us, Georges wants to retain a certain number of iterations as the minimum length of the steady state (their `k`), so we need to use the same value for `k` for both methods. If we say "Might Georges find a steady state on iteration 1750", then I think we need to ask the same question of our method, otherwise the comparison is skewed.

**Comment by ltratt** [comment]
Good point.

**Comment by ltratt** [comment]
So I think we can merge this after squashing?

**Comment by snim2** [comment]
Squashed!

**Pull request merged**